### DISCUSSION TITLE
How to access data from another plugin in a hook?

### DISCUSSION THREAD

**Discussion by silbe** [pytest-dev/pytest#10129] [Q&A]
I've written a plugin that streams test results into a file and provides fixtures that can be used by both fixtures and tests to include structured data in the test results file. The fixtures store the data on the [stash](https://docs.pytest.org/en/stable/how-to/writing_hook_functions.html#storing-data-on-items-across-hook-functions) of the Item. A `pytest_runtest_makereport()` hook copies (a reference to) the data from the Item stash to the Report stash. In the `pytest_make_collect_report()` both the test result and the structured data on the Report stash are used to output the test results.

This works fine so far. Now I'd like to factor out the (pretty generic) structured data fixtures into a separate plugin so they can be used independently of the specific output format of the test result plugin. If I were using only fixtures, that would be pretty easy; I'd just request the fixtures of the structured data plugin from the fixtures in the test result plugin and be done with it. However I need to access the data from within hooks, not fixtures.

I have a couple of ideas how to achieve that but none of them feels like a good solution:
- Instead of using `pytest.StashKey`, use a fixed string and let the other plugins hard-code that string.
- Provide getter functions in my plugin and let the other plugins get a reference to my plugin using the plugin manager and invoke the functions on that reference; alternatively just look up the `StashKey` on the plugin reference. (Does using the plugin manager to access functions or variables in a plugin even work?)
- Add an autouse fixture to each plugin that adds a reference to the structured data to a per-plugin Item stash entry; use the per-plugin stash entry to access the structured data from hooks.

Is there a better approach?

**Comment by nicoddemus** [comment]
Hi @silbe,

I think the solution is a matter of taste, but I would make the StashKey part of the public API of the plugin, and the other plugins can access the value in the stash using the key directly.

**Reply by silbe** [reply]
How would you do that? How does one access public API of another plugin?

**Reply by nicoddemus** [reply]
Oh just import it directly:

```python
import my_shared_plugin
my_shared_plugin.Key
```

Something like that.

**Reply by silbe** [reply]
Hmm, this results in a warning that assertions cannot be rewritten for the shared plugin:
```
/usr/lib/python3/dist-packages/_pytest/config/__init__.py:663
  /usr/lib/python3/dist-packages/_pytest/config/__init__.py:663: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: pytest_structured_result
    self.import_plugin(import_spec)
```

I guess I could invoke `pytest.register_assert_rewrite('pytest_structured_result')` before importing the plugin, but that adds magic boiler-plate code for all users of the plugin and will also yield a PEP8 warning (again for all users of the plugin) because the import is done after a function call.

FWIW the shared plugin currently lives directly in the repository that contains the tests. It's going to be moved to a submodule later and I'll probably also support installing it. The users of the plugin should not need to care about where it's located, as long as it can be loaded by pytest.

**Comment by nicoddemus** [comment]
> I guess I could invoke pytest.register_assert_rewrite('pytest_structured_result') before importing the plugin, but that adds magic boiler-plate code for all users of the plugin and will also yield a PEP8 warning (again for all users of the plugin) because the import is done after a function call.

Alternatively you can instruct users to only import your plugin locally inside functions and methods which need to access the key.
